File: sea_battle.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

#define BOARD_SIZE 5
#define NUM_SHIPS 3

typedef struct {
    int x;
    int y;
} Coordinates;

typedef struct {
    char board[BOARD_SIZE][BOARD_SIZE];
} GameBoard;

void initialize_board(GameBoard *board) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            board->board[i][j] = '-';
        }
    }
}

void print_board(GameBoard *board) {
    printf("  ");
    for (int i = 0; i < BOARD_SIZE; i++) {
        printf("%d ", i);
    }
    printf("\n");

    for (int i = 0; i < BOARD_SIZE; i++) {
        printf("%d ", i);
        for (int j = 0; j < BOARD_SIZE; j++) {
            printf("%c ", board->board[i][j]);
        }
        printf("\n");
    }
}

bool is_valid_coordinate(int x, int y) {
    return (x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE);
}

bool is_valid_placement(GameBoard *board, Coordinates coord) {
    return (board->board[coord.x][coord.y] == '-');
}

void place_ships(GameBoard *board, int num_ships) {
    srand(time(NULL));

    for (int ship = 0; ship < num_ships; ship++) {
        Coordinates coord;
        do {
            coord.x = rand() % BOARD_SIZE;
            coord.y = rand() % BOARD_SIZE;
        } while (!is_valid_placement(board, coord));

        board->board[coord.x][coord.y] = 'S';
    }
}

bool is_game_over(GameBoard *board) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            if (board->board[i][j] == 'S') {
                return false;
            }
        }
    }
    return true;
}

void player_turn(GameBoard *computer_board) {
    int x, y;
    do {
        printf("Enter coordinates to attack (x y): ");
        scanf("%d %d", &x, &y);
    } while (!is_valid_coordinate(x, y));

    if (computer_board->board[x][y] == 'S') {
        printf("Hit!\n");
        computer_board->board[x][y] = 'X';
    } else if (computer_board->board[x][y] == '-') {
        printf("Miss!\n");
        computer_board->board[x][y] = 'O';
    } else {
        printf("You've already fired at this location. Try again.\n");
    }
}

void computer_turn(GameBoard *player_board) {
    int x, y;
    do {
        x = rand() % BOARD_SIZE;
        y = rand() % BOARD_SIZE;
    } while (!is_valid_coordinate(x, y));

    if (player_board->board[x][y] == 'S') {
        printf("Computer hits at (%d, %d)! ", x, y);
        player_board->board[x][y] = 'X';
    } else if (player_board->board[x][y] == '-') {
        printf("Computer misses at (%d, %d). ", x, y);
        player_board->board[x][y] = 'O';
    } else {
        printf("Computer already fired at (%d, %d). ", x, y);
    }
}

int main() {
    GameBoard player_board, computer_board;
    int num_ships_remaining = NUM_SHIPS;

    initialize_board(&player_board);
    initialize_board(&computer_board);

    place_ships(&player_board, NUM_SHIPS);
    place_ships(&computer_board, NUM_SHIPS);

    printf("Welcome to Sea Battle!\n");
    printf("Try to sink the enemy ships!\n");

    while (!is_game_over(&computer_board) && !is_game_over(&player_board)) {
        printf("\nYour board:\n");
        print_board(&player_board);
        printf("\nComputer's board:\n");
        print_board(&computer_board);

        player_turn(&computer_board);
        if (is_game_over(&computer_board)) {
            printf("You win! All enemy ships sunk!\n");
            break;
        }

        computer_turn(&player_board);
        if (is_game_over(&player_board)) {
            printf("Computer wins! All your ships sunk!\n");
            break;
        }
    }

    return 0;
}

File: python_code/sea-battle.py
import random
from flask import Flask, render_template, request, redirect, url_for

app = Flask(__name__)

BOARD_SIZE = 5
NUM_SHIPS = 3

class GameBoard:
    def __init__(self):
        self.board = [['-' for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]

    def print_board(self):
        return "\n".join(" ".join(row) for row in self.board)

def is_valid_coordinate(x, y):
    return 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE

def place_ships(board, num_ships):
    for _ in range(num_ships):
        while True:
            x, y = random.randint(0, BOARD_SIZE - 1), random.randint(0, BOARD_SIZE - 1)
            if board[x][y] != 'S':
                break
        board[x][y] = 'S'

def player_turn(computer_board, x, y):
    if computer_board.board[x][y] == 'S':
        computer_board.board[x][y] = 'X'
        return "Hit!"
    elif computer_board.board[x][y] == '-':
        computer_board.board[x][y] = 'O'
        return "Miss!"
    else:
        return "You've already fired at this location. Try again."

def computer_turn(player_board):
    while True:
        x, y = random.randint(0, BOARD_SIZE - 1), random.randint(0, BOARD_SIZE - 1)
        if player_board.board[x][y] == '-':
            break

    if player_board.board[x][y] == 'S':
        player_board.board[x][y] = 'X'
        return f"Computer hits at ({x}, {y})!"
    elif player_board.board[x][y] == '-':
        player_board.board[x][y] = 'O'
        return f"Computer misses at ({x}, {y})."
    else:
        return f"Computer already fired at ({x}, {y})."

def is_game_over(board):
    for row in board.board:
        if 'S' in row:
            return False
    return True

player_board = GameBoard()
computer_board = GameBoard()
place_ships(player_board.board, NUM_SHIPS)
place_ships(computer_board.board, NUM_SHIPS)

@app.route('/')
def index():
    return render_template('index.html')

@app.route('/attack', methods=['POST'])
def attack():
    if 'x' in request.form and 'y' in request.form:
        x = int(request.form['x'])
        y = int(request.form['y'])
        if is_valid_coordinate(x, y):
            message = player_turn(computer_board, x, y)
            if is_game_over(computer_board):
                return render_template('result.html', message=message, winner="Player")
            message += "\n" + computer_turn(player_board)
            if is_game_over(player_board):
                return render_template('result.html', message=message, winner="Computer")
            return render_template('game.html', player_board=player_board.print_board(), message=message)
    return redirect(url_for('index'))

if __name__ == "__main__":
    app.run(debug=True)
